File: 0142_Linked_List_Cycle_II/0142_Linked_List_Cycle_II.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */


/**
 *	微信号: VINO
 *	先判断是否成环，记录快慢指针第一次相遇时将慢指针归head处 
 *  快慢指针在环内一定会相遇且相遇点只能是环入口，否则俩指针速度一样，在环内永不可能相遇
 *
 */

public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode slow = head, fast = head;
        while(fast != null && fast.next != null)
        {
        	slow = slow.next;
        	fast = fast.next.next;
        	if(slow == fast)
        		break;
        }
        
        if(fast == null || fast.next == null)
        	return null;
        
        slow = head;//慢指针重新归位头指针处
        while(slow != fast)
        {
        	slow = slow.next;
        	fast = fast.next;
        }
        return slow;
    }
}

File: problems/0142_Linked_List_Cycle_II/0142_Linked_List_Cycle_II.cpp
/*
  题目：
  给定一个链表，返回链表开始入环的第一个节点。 如果链表无环，则返回 null。
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* slow=head;
        ListNode* fast=head;
        ListNode* encounter=NULL;
        bool existCycle=false;
        while(fast && fast->next && fast->next->next){
            slow=slow->next;
            fast=fast->next->next;
            if(slow==fast){
                encounter=fast;
                existCycle=true;
                break;
            }
        }
        
        if( existCycle ){
            ListNode* p=head;
            while(p!=encounter){
                p=p->next;
                encounter=encounter->next;
            }
            return p;
        }
        
        return NULL;
    }
};
